src/Pos/Explorer/BListener.hs

```diff
@@ -0,0 +1,255 @@
+import           Pos.Core                     (HeaderHash, difficultyL,
+                                               epochIndexL, getChainDifficulty,
+                                               headerHash)
+import           Pos.DB.BatchOp               (SomeBatchOp (..))
+import           Pos.DB.Class                 (MonadDBRead, MonadRealDB)
+import           Pos.Explorer.DB              (Page, Epoch)
+import qualified Pos.Explorer.DB              as DB
+import           Pos.Ssc.Class.Helpers        (SscHelpersClass)
+import           Pos.Util.Chrono              (NE, NewestFirst (..),
+                                               OldestFirst (..))
+
+----------------------------------------------------------------------------
+-- Declarations
+----------------------------------------------------------------------------
+
+data ExplorerBListenerTag
+
+type ExplorerBListener = Ether.TaggedTrans ExplorerBListenerTag IdentityT
+
+-- Kind of... Empty and devoid of meaning...
+runExplorerBListener :: ExplorerBListener m a -> m a
+runExplorerBListener = coerce
+
+-- Explorer implementation for usual node. Combines the operations.
+instance ( MonadRealDB m
+         , MonadDBRead m
+         , MonadMockable m
+         , WithLogger m
+         )
+         => MonadBListener (ExplorerBListener m) where
+    onApplyBlocks     blunds = do
+        epochBlocks <- onApplyEpochBlocksExplorer blunds
+        pageBlocks  <- onApplyPageBlocksExplorer blunds
+        pure $ SomeBatchOp [epochBlocks, pageBlocks]
+
+    onRollbackBlocks  blunds = do
+        epochBlocks <- onRollbackEpochBlocksExplorer blunds
+        pageBlocks  <- onRollbackPageBlocksExplorer blunds
+        pure $ SomeBatchOp [epochBlocks, pageBlocks]
+
+-- onGeneralEpochBlocksExplorer is viable, since the only thing that needs 
+-- to be defined are reversed operations (adding and removing)
+
+----------------------------------------------------------------------------
+-- Apply
+----------------------------------------------------------------------------
+
+-- For @EpochBlocks@
+onApplyEpochBlocksExplorer
+    :: forall ssc m .
+    ( SscHelpersClass ssc
+    , WithLogger m
+    , MonadRealDB m
+    , MonadDBRead m
+    )
+    => OldestFirst NE (Blund ssc) -> m SomeBatchOp
+onApplyEpochBlocksExplorer blunds = do
+    -- Get the epoch and the attached block @HeaderHash@es.
+    let epoch    = epochBlocks ^. _1
+    let blockHHs = epochBlocks ^. _2
+    
+    -- Get all blocks from the current epoch (so we can remove everything and
+    -- add the (Epoch, [HeaderHash]) pair back to the database).
+    mBlocksHHs <- DB.getEpochBlocks epoch
+    let blocksHHsOrEmpty = fromMaybe mempty mBlocksHHs
+    let uniqueHHs = ordNub $ blocksHHsOrEmpty <> blockHHs
+
+    -- In the end make sure we place this under @SomeBatchOp@ in order to
+    -- preserve atomicity.
+    pure $ SomeBatchOp $ DB.PutEpochBlocks epoch uniqueHHs
+  where
+    epochBlocks = minEpochBlocks blocksNE
```

Comment:
`minEpochBlocks` seems useless here, because we know that all blocks are from the same epoch (which is false for rollback, btw).